File: crates/hooviestar-engine/src/project.rs
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use uuid::Uuid;

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ProjectV1 {
    pub version: u32,
    pub output: OutputConfig,
    pub sources: Vec<Source>,
    pub scenes: Vec<Scene>,
    pub active_scene_id: Uuid,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutputConfig {
    pub width: u32,
    pub height: u32,
    pub fps: u32,
    pub background: String,
}
impl Default for OutputConfig {
    fn default() -> Self {
        Self {
            width: 1280,
            height: 720,
            fps: 30,
            background: "#101418".into(),
        }
    }
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct WindowBinding {
    pub process_path: String,
    pub window_title: String,
}
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DisplayBinding {
    pub adapter_luid: String,
    pub output_id: u32,
}
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AudioSessionBinding {
    pub process_path: String,
    pub session_grouping_id: String,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TextAlign {
    Left,
    Center,
    Right,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(
    tag = "type",
    rename_all = "snake_case",
    rename_all_fields = "camelCase"
)]
pub enum Source {
    Window {
        id: Uuid,
        name: String,
        binding: WindowBinding,
    },
    Display {
        id: Uuid,
        name: String,
        binding: DisplayBinding,
    },
    Image {
        id: Uuid,
        name: String,
        path: String,
    },
    Text {
        id: Uuid,
        name: String,
        text: String,
        font_family: String,
        font_size_px: f32,
        font_weight: u16,
        color: String,
        background_color: String,
        align: TextAlign,
    },
    Media {
        id: Uuid,
        name: String,
        path: String,
        #[serde(rename = "loop")]
        looped: bool,
        continue_when_hidden: bool,
        restart_on_show: bool,
        volume: f32,
        muted: bool,
    },
    ApplicationAudio {
        id: Uuid,
        name: String,
        binding: AudioSessionBinding,
        volume: f32,
        muted: bool,
    },
}
impl Source {
    pub fn id(&self) -> Uuid {
        match self {
            Self::Window { id, .. }
            | Self::Display { id, .. }
            | Self::Image { id, .. }
            | Self::Text { id, .. }
            | Self::Media { id, .. }
            | Self::ApplicationAudio { id, .. } => *id,
        }
    }
    pub fn name(&self) -> &str {
        match self {
            Self::Window { name, .. }
            | Self::Display { name, .. }
            | Self::Image { name, .. }
            | Self::Text { name, .. }
            | Self::Media { name, .. }
            | Self::ApplicationAudio { name, .. } => name,
        }
    }
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Scene {
    pub id: Uuid,
    pub name: String,
    pub hotkey: Option<String>,
    pub items: Vec<SceneItem>,
}
#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SceneItem {
    pub id: Uuid,
    pub source_id: Uuid,
    pub visible: bool,
    pub locked: bool,
    pub transform: Transform,
}
#[derive(Clone, Copy, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Transform {
    pub x: f32,
    pub y: f32,
    pub width: f32,
    pub height: f32,
    pub rotation_degrees: f32,
    pub crop_top: f32,
    pub crop_right: f32,
    pub crop_bottom: f32,
    pub crop_left: f32,
    pub opacity: f32,
}
impl Default for Transform {
    fn default() -> Self {
        Self {
            x: 0.0,
            y: 0.0,
            width: 1280.0,
            height: 720.0,
            rotation_degrees: 0.0,
            crop_top: 0.0,
            crop_right: 0.0,
            crop_bottom: 0.0,
            crop_left: 0.0,
            opacity: 1.0,
        }
    }
}

impl ProjectV1 {
    pub fn empty() -> Self {
        let game = Uuid::new_v4();
        let video = Uuid::new_v4();
        let both = Uuid::new_v4();
        Self {
            version: 1,
            output: OutputConfig::default(),
            sources: vec![],
            scenes: vec![
                Scene {
                    id: game,
                    name: "Spiel".into(),
                    hotkey: Some("Ctrl+Alt+1".into()),
                    items: vec![],
                },
                Scene {
                    id: video,
                    name: "Video".into(),
                    hotkey: Some("Ctrl+Alt+2".into()),
                    items: vec![],
                },
                Scene {
                    id: both,
                    name: "Beides".into(),
                    hotkey: Some("Ctrl+Alt+3".into()),
                    items: vec![],
                },
            ],
            active_scene_id: game,
        }
    }
    pub fn validate(&self) -> Result<(), String> {
        if self.version != 1 {
            return Err("unsupported project version".into());
        }
        if !matches!(
            (self.output.width, self.output.height),
            (1280, 720) | (1920, 1080)
        ) || !matches!(self.output.fps, 30 | 60)
        {
            return Err("unsupported output preset".into());
        }
        if !is_color(&self.output.background) {
            return Err("invalid background color".into());
        }
        if self.scenes.is_empty() {
            return Err("project requires a scene".into());
        }
        let mut all_ids = HashSet::new();
        let mut source_ids = HashSet::new();
        for source in &self.sources {
            if !source_ids.insert(source.id()) || !all_ids.insert(source.id()) {
                return Err("duplicate source id".into());
            }
            if source.name().trim().is_empty() {
                return Err("source name is empty".into());
            }
            if let Source::Text {
                color,
                background_color,
                ..
            } = source
                && (!is_color(color) || !is_color(background_color))
            {
                return Err("invalid text color".into());
            }
            let font_size_px = match source {
                Source::Text { font_size_px, .. } => Some(*font_size_px),
                _ => None,
            };
            if font_size_px.is_some_and(|size| !size.is_finite() || size <= 0.0 || size > 4096.0) {
                return Err("invalid text font size".into());
            }
            let volume = match source {
                Source::Media { volume, .. } | Source::ApplicationAudio { volume, .. } => {
                    Some(*volume)
                }
                _ => None,
            };
            if volume.is_some_and(|volume| !volume.is_finite() || !(0.0..=1.0).contains(&volume)) {
                return Err("invalid source volume".into());
            }
        }
        let mut scene_ids = HashSet::new();
        let mut hotkeys = HashSet::new();
        let mut item_ids = HashSet::new();
        for scene in &self.scenes {
            let mut scene_source_ids = HashSet::new();
            if scene.name.trim().is_empty() {
                return Err("scene name is empty".into());
            }
            if !scene_ids.insert(scene.id) || !all_ids.insert(scene.id) {
                return Err("duplicate scene id".into());
            }
            if let Some(hotkey) = &scene.hotkey {
                if hotkey.trim().is_empty() {
                    return Err("invalid scene hotkey".into());
                }
                if !hotkeys.insert(hotkey.to_ascii_lowercase()) {
                    return Err("duplicate scene hotkey".into());
                }
            }
            if scene.items.len() > 128 {
                return Err("scene has too many items".into());
            }
            for item in &scene.items {
                if !item_ids.insert(item.id) || !all_ids.insert(item.id) {
                    return Err("duplicate item id".into());
                }
                if !source_ids.contains(&item.source_id) {
                    return Err("scene item source is missing".into());
                }
                if !scene_source_ids.insert(item.source_id) {
                    return Err("source appears more than once in scene".into());
                }
                let t = item.transform;
                if ![
                    t.x,
                    t.y,
                    t.width,
                    t.height,
                    t.rotation_degrees,
                    t.crop_top,
                    t.crop_right,
                    t.crop_bottom,
                    t.crop_left,
                    t.opacity,
                ]
                .iter()
                .all(|v| v.is_finite())
                    || t.width <= 0.0
                    || t.height <= 0.0
                    || t.width > 8192.0
                    || t.height > 8192.0
                    || t.crop_top < 0.0
                    || t.crop_right < 0.0
                    || t.crop_bottom < 0.0
                    || t.crop_left < 0.0
                    || t.crop_left + t.crop_right >= t.width
                    || t.crop_top + t.crop_bottom >= t.height
                    || !(0.0..=1.0).contains(&t.opacity)
                {
                    return Err("invalid transform".into());
                }
            }
        }
        if !scene_ids.contains(&self.active_scene_id) {
            return Err("active scene is missing".into());
        }
        Ok(())
    }
}

fn is_color(value: &str) -> bool {
    value.len() == 7 && value.starts_with('#') && value[1..].bytes().all(|b| b.is_ascii_hexdigit())
}
#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_project_has_three_hotkey_scenes() {
        let project = ProjectV1::empty();
        assert_eq!(
            project
                .scenes
                .iter()
                .map(|scene| (scene.name.as_str(), scene.hotkey.as_deref()))
                .collect::<Vec<_>>(),
            vec![
                ("Spiel", Some("Ctrl+Alt+1")),
                ("Video", Some("Ctrl+Alt+2")),
                ("Beides", Some("Ctrl+Alt+3")),
            ]
        );
        project.validate().unwrap();
    }

    #[test]
    fn duplicate_hotkey_is_rejected() {
        let mut project = ProjectV1::empty();
        project.scenes[1].hotkey = project.scenes[0].hotkey.clone();
        assert_eq!(project.validate(), Err("duplicate scene hotkey".into()));
    }

    #[test]
    fn qualified_output_dimensions_and_frame_rates_are_independent() {
        for (width, height) in [(1280, 720), (1920, 1080)] {
            for fps in [30, 60] {
                let mut project = ProjectV1::empty();
                project.output = OutputConfig {
                    width,
                    height,
                    fps,
                    background: "#aBc123".into(),
                };
                project.validate().unwrap();
            }
        }
        for output in [
            OutputConfig {
                width: 2560,
                height: 1440,
                fps: 60,
                background: "#101418".into(),
            },
            OutputConfig {
                width: 1280,
                height: 720,
                fps: 24,
                background: "#101418".into(),
            },
        ] {
            let mut project = ProjectV1::empty();
            project.output = output;
            assert_eq!(project.validate(), Err("unsupported output preset".into()));
        }
    }

    #[test]
    fn invalid_audio_volume_is_rejected() {
        let mut project = ProjectV1::empty();
        project.sources.push(Source::ApplicationAudio {
            id: Uuid::new_v4(),
            name: "Audio".into(),
            binding: AudioSessionBinding {
                process_path: "/usr/bin/game".into(),
                session_grouping_id: "game".into(),
            },
            volume: f32::NAN,
            muted: false,
        });
        assert_eq!(project.validate(), Err("invalid source volume".into()));
    }

    #[test]
    fn blank_source_name_is_rejected() {
        let mut project = ProjectV1::empty();
        project.sources.push(Source::ApplicationAudio {
            id: Uuid::new_v4(),
            name: "   ".into(),
            binding: AudioSessionBinding {
                process_path: "/usr/bin/game".into(),
                session_grouping_id: "game".into(),
            },
            volume: 1.0,
            muted: false,
        });
        assert_eq!(project.validate(), Err("source name is empty".into()));
        match project.sources.first_mut() {
            Some(Source::ApplicationAudio { name, .. }) => *name = "Game-Audio".into(),
            _ => unreachable!("test pushed an ApplicationAudio source"),
        }
        project.validate().unwrap();
    }

    #[test]
    fn blank_hotkey_is_rejected() {
        let mut project = ProjectV1::empty();
        project.scenes[0].hotkey = Some("   ".into());
        assert_eq!(project.validate(), Err("invalid scene hotkey".into()));
        project.scenes[0].hotkey = Some("Ctrl+Alt+9".into());
        project.validate().unwrap();
    }

    #[test]
    fn duplicate_source_in_one_scene_is_rejected() {
        let mut project = ProjectV1::empty();
        let source_id = Uuid::new_v4();
        project.sources.push(Source::Image {
            id: source_id,
            name: "Bild".into(),
            path: "/tmp/image.png".into(),
        });
        project.scenes[0].items.extend([
            SceneItem {
                id: Uuid::new_v4(),
                source_id,
                visible: true,
                locked: false,
                transform: Transform::default(),
            },
            SceneItem {
                id: Uuid::new_v4(),
                source_id,
                visible: true,
                locked: false,
                transform: Transform::default(),
            },
        ]);
        assert_eq!(
            project.validate(),
            Err("source appears more than once in scene".into())
        );
    }
}

File: src/types/project.ts
/**
 * Persistierte Projektverträge – exakte Spiegelbilder der Rust-Typen aus
 * `hooviestar-engine/src/project.rs`.
 *
 * Konvention (serde): Tag-Feld `type`, Variantennamen snake_case,
 * Felder camelCase. Diese Datei beschreibt ausschließlich Zustand, den die
 * Engine autoritativ besitzt – das UI erfindet niemals persistierten Zustand.
 */

/** UUID v4 als String, wie von Rust `uuid::Uuid` serialisiert. */
export type Uuid = string;

/** sRGB-Hexfarbe inklusive `#`, z. B. `#101418`. */
export type HexColor = string;

export interface OutputConfig {
  width: number;
  height: number;
  fps: number;
  background: HexColor;
}

export interface Transform {
  /** Linke Kante in Ausgabepixeln. */
  x: number;
  /** Obere Kante in Ausgabepixeln. */
  y: number;
  width: number;
  height: number;
  /** Drehung in Grad, gegen den Uhrzeigersinn. */
  rotationDegrees: number;
  cropTop: number;
  cropRight: number;
  cropBottom: number;
  cropLeft: number;
  /** 0.0 ..= 1.0 */
  opacity: number;
}

/** Bindung einer Fensterquelle: kanonischer Prozesspfad + exakter Titel. */
export interface WindowBinding {
  processPath: string;
  windowTitle: string;
}

/** Bindung einer Monitorquelle: Adapter-LUID + Output-ID. */
export interface DisplayBinding {
  adapterLuid: string;
  outputId: number;
}

/** Bindung einer Audiositzung: Prozesspfad + Sitzungs-Gruppierungskennung. */
export interface AudioSessionBinding {
  processPath: string;
  sessionGroupingId: string;
}

interface SourceCommon {
  id: Uuid;
  name: string;
}

/** Fensterquelle (Windows Graphics Capture, Borderless/DWM). */
export interface WindowSource extends SourceCommon {
  type: "window";
  binding: WindowBinding;
}

/** Monitorquelle (Desktop Duplication, einziger Pfad für Exclusive Fullscreen). */
export interface DisplaySource extends SourceCommon {
  type: "display";
  binding: DisplayBinding;
}

/** Einmalig dekodierte Bilddatei als immutable GPU-Textur. */
export interface ImageSource extends SourceCommon {
  type: "image";
  /** Kanonischer absoluter Pfad. */
  path: string;
}

export type TextAlign = "left" | "center" | "right";

export interface TextSource extends SourceCommon {
  type: "text";
  text: string;
  fontFamily: string;
  fontSizePx: number;
  fontWeight: number;
  color: HexColor;
  backgroundColor: HexColor;
  align: TextAlign;
}

export interface MediaSource extends SourceCommon {
  type: "media";
  /** Kanonischer absoluter Pfad (H.264/AAC-MP4, MP3, WAV garantiert). */
  path: string;
  loop: boolean;
  continueWhenHidden: boolean;
  restartOnShow: boolean;
  /** 0.0 ..= 1.0 */
  volume: number;
  muted: boolean;
}

/** Nur-Audio-Quelle; wird gemischt, hat aber keinen sichtbaren Layer. */
export interface ApplicationAudioSource extends SourceCommon {
  type: "application_audio";
  binding: AudioSessionBinding;
  /** 0.0 ..= 1.0 */
  volume: number;
  muted: boolean;
}

export type Source =
  | WindowSource
  | DisplaySource
  | ImageSource
  | TextSource
  | MediaSource
  | ApplicationAudioSource;

export type SourceType = Source["type"];

export interface SceneItem {
  id: Uuid;
  sourceId: Uuid;
  visible: boolean;
  locked: boolean;
  transform: Transform;
}

export interface Scene {
  id: Uuid;
  name: string;
  /** Global registrierter Hotkey wie `Ctrl+Alt+1`; `null` = ungebunden. */
  hotkey: string | null;
  /**
   * Zeichenreihenfolge: früherer Eintrag unten, späterer Eintrag oben.
   */
  items: SceneItem[];
}

/** Einzige persistierte Projektwurzel (`version` ist literal `1`). */
export interface ProjectV1 {
  version: 1;
  output: OutputConfig;
  sources: Source[];
  scenes: Scene[];
  activeSceneId: Uuid;
}

/* ------------------------------------------------------------------ */
/* Laufzeit-Wächter (Validierung externer Daten, u. a. Fixtures)       */
/* ------------------------------------------------------------------ */

export const SOURCE_TYPES: readonly SourceType[] = [
  "window",
  "display",
  "image",
  "text",
  "media",
  "application_audio",
];

export function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function str(value: unknown): value is string {
  return typeof value === "string";
}

function num(value: unknown): value is number {
  return typeof value === "number" && Number.isFinite(value);
}

function bool(value: unknown): value is boolean {
  return typeof value === "boolean";
}

/** Canonical hyphenated UUID accepted by Rust's `uuid::Uuid` boundary. */
export function isUuid(value: unknown): value is Uuid {
  return (
    typeof value === "string" &&
    /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(value)
  );
}

function uint(value: unknown, max: number): value is number {
  return num(value) && Number.isInteger(value) && value >= 0 && value <= max;
}

function color(value: unknown): value is HexColor {
  return typeof value === "string" && /^#[0-9a-fA-F]{6}$/.test(value);
}

export function parseTransform(value: unknown): Transform | null {
  if (!isRecord(value)) return null;
  const keys = [
    "x",
    "y",
    "width",
    "height",
    "rotationDegrees",
    "cropTop",
    "cropRight",
    "cropBottom",
    "cropLeft",
    "opacity",
  ] as const;
  for (const k of keys) {
    if (!num(value[k])) return null;
  }
  const opacity = value.opacity as number;
  if (opacity < 0 || opacity > 1) return null;
  // project.rs:310-320 – zusätzlich zu den endlichen Feldern (num oben):
  // Breite/Höhe größer 0 und höchstens 8192, Crops nicht negativ,
  // Cropsummen bleiben unterhalb der Abmessungen.
  const width = value.width as number;
  const height = value.height as number;
  const cropTop = value.cropTop as number;
  const cropRight = value.cropRight as number;
  const cropBottom = value.cropBottom as number;
  const cropLeft = value.cropLeft as number;
  if (
    width <= 0 ||
    height <= 0 ||
    width > 8192 ||
    height > 8192 ||
    cropTop < 0 ||
    cropRight < 0 ||
    cropBottom < 0 ||
    cropLeft < 0 ||
    cropLeft + cropRight >= width ||
    cropTop + cropBottom >= height
  ) {
    return null;
  }
  return {
    x: value.x as number,
    y: value.y as number,
    width: value.width as number,
    height: value.height as number,
    rotationDegrees: value.rotationDegrees as number,
    cropTop: value.cropTop as number,
    cropRight: value.cropRight as number,
    cropBottom: value.cropBottom as number,
    cropLeft: value.cropLeft as number,
    opacity,
  };
}

export function parseSource(value: unknown): Source | null {
  if (!isRecord(value)) return null;
  const type = value.type;
  if (!SOURCE_TYPES.includes(type as SourceType)) return null;
  if (!isUuid(value.id) || !str(value.name)) return null;
  const id = value.id as string;
  const name = value.name as string;
  // project.rs:238-240 – leere Quellnamen (nach Trimmen) sind ungültig.
  if (name.trim().length === 0) return null;
  switch (type) {
    case "window": {
      const b = value.binding;
      if (!isRecord(b) || !str(b.processPath) || !str(b.windowTitle)) return null;
      return { type, id, name, binding: { processPath: b.processPath, windowTitle: b.windowTitle } };
    }
    case "display": {
      const b = value.binding;
      if (!isRecord(b) || !str(b.adapterLuid) || !uint(b.outputId, 0xffff_ffff)) return null;
      return { type, id, name, binding: { adapterLuid: b.adapterLuid, outputId: b.outputId } };
    }
    case "image":
      return str(value.path) ? { type, id, name, path: value.path } : null;
    case "text": {
      if (
        !str(value.text) ||
        !str(value.fontFamily) ||
        !num(value.fontSizePx) ||
        !uint(value.fontWeight, 0xffff) ||
        !color(value.color) ||
        !color(value.backgroundColor)
      ) {
        return null;
      }
      const align = value.align;
      if (align !== "left" && align !== "center" && align !== "right") return null;
      // project.rs:254-256 – Schriftgröße endlich (num oben), > 0, <= 4096.
      const fontSizePx = value.fontSizePx as number;
      if (fontSizePx <= 0 || fontSizePx > 4096) return null;
      return {
        type,
        id,
        name,
        text: value.text,
        fontFamily: value.fontFamily,
        fontSizePx: value.fontSizePx,
        fontWeight: value.fontWeight,
        color: value.color,
        backgroundColor: value.backgroundColor,
        align,
      };
    }
    case "media": {
      if (
        !str(value.path) ||
        !bool(value.loop) ||
        !bool(value.continueWhenHidden) ||
        !bool(value.restartOnShow) ||
        !num(value.volume) || value.volume < 0 || value.volume > 1 ||
        !bool(value.muted)
      ) {
        return null;
      }
      return {
        type,
        id,
        name,
        path: value.path,
        loop: value.loop,
        continueWhenHidden: value.continueWhenHidden,
        restartOnShow: value.restartOnShow,
        volume: value.volume,
        muted: value.muted,
      };
    }
    case "application_audio": {
      const b = value.binding;
      if (!isRecord(b) || !str(b.processPath) || !str(b.sessionGroupingId)) return null;
      if (!num(value.volume) || value.volume < 0 || value.volume > 1 || !bool(value.muted)) return null;
      return {
        type,
        id,
        name,
        binding: { processPath: b.processPath, sessionGroupingId: b.sessionGroupingId },
        volume: value.volume,
        muted: value.muted,
      };
    }
  }
  return null;
}

export function parseSceneItem(value: unknown): SceneItem | null {
  if (!isRecord(value)) return null;
  const transform = parseTransform(value.transform);
  if (!isUuid(value.id) || !isUuid(value.sourceId) || !bool(value.visible) || !bool(value.locked) || !transform) {
    return null;
  }
  return {
    id: value.id as string,
    sourceId: value.sourceId as string,
    visible: value.visible as boolean,
    locked: value.locked as boolean,
    transform,
  };
}

export function parseScene(value: unknown): Scene | null {
  if (!isRecord(value) || !isUuid(value.id) || !str(value.name)) return null;
  // project.rs:271-273 – leere Szenennamen (nach Trimmen) sind ungültig.
  if ((value.name as string).trim().length === 0) return null;
  // project.rs:141/serde – abwesender Schlüssel zählt wie null als kein
  // Hotkey; project.rs:277-280 – leerer Hotkey (nach Trimmen) ungültig.
  if (value.hotkey !== null && value.hotkey !== undefined) {
    if (!str(value.hotkey) || (value.hotkey as string).trim().length === 0) return null;
  }
  if (!Array.isArray(value.items)) return null;
  // project.rs:285-287 – höchstens 128 Items pro Szene.
  if (value.items.length > 128) return null;
  const items: SceneItem[] = [];
  for (const raw of value.items) {
    const item = parseSceneItem(raw);
    if (!item) return null;
    items.push(item);
  }
  return {
    id: value.id as string,
    name: value.name as string,
    hotkey: (value.hotkey ?? null) as string | null,
    items,
  };
}

export function parseOutputConfig(value: unknown): OutputConfig | null {
  if (!isRecord(value) || !num(value.width) || !num(value.height) || !num(value.fps) || !color(value.background)) {
    return null;
  }
  // project.rs:220-225 – nur die freigegebenen Presets (1280,720,30)
  // und (1920,1080,60) sind gültig.
  const outWidth = value.width as number;
  const outHeight = value.height as number;
  const outFps = value.fps as number;
  if (
    !(outWidth === 1280 && outHeight === 720 && outFps === 30) &&
    !(outWidth === 1920 && outHeight === 1080 && outFps === 60)
  ) {
    return null;
  }
  return {
    width: value.width as number,
    height: value.height as number,
    fps: value.fps as number,
    background: value.background as HexColor,
  };
}

/** Validiert ein vollständiges `ProjectV1` inklusive Referenzintegrität. */
export function parseProjectV1(value: unknown): ProjectV1 | null {
  if (!isRecord(value)) return null;
  if (value.version !== 1) return null;
  const output = parseOutputConfig(value.output);
  if (!output) return null;
  if (!Array.isArray(value.sources) || !Array.isArray(value.scenes)) return null;
  if (!isUuid(value.activeSceneId)) return null;
  const sources: Source[] = [];
  const sourceIds = new Set<string>();
  // project.rs:232/267-269 – ein ID-Namespace über Quellen, Szenen und Items.
  const allIds = new Set<string>();
  for (const raw of value.sources) {
    const source = parseSource(raw);
    if (!source || sourceIds.has(source.id)) return null;
    sources.push(source);
    sourceIds.add(source.id);
    allIds.add(source.id);
  }
  const scenes: Scene[] = [];
  const sceneIds = new Set<string>();
  const hotkeys = new Set<string>();
  const itemIds = new Set<string>();
  for (const raw of value.scenes) {
    const scene = parseScene(raw);
    if (!scene) return null;
    // project.rs:274-276 – doppelte Szenen-ID oder Kollision über Arten.
    if (sceneIds.has(scene.id) || allIds.has(scene.id)) return null;
    sceneIds.add(scene.id);
    allIds.add(scene.id);
    if (scene.hotkey !== null) {
      // project.rs:281-283 – Hotkeys case-insensitive projektweit eindeutig.
      const normalized = scene.hotkey.toLowerCase();
      if (hotkeys.has(normalized)) return null;
      hotkeys.add(normalized);
    }
    const sceneSourceIds = new Set<string>();
    for (const item of scene.items) {
      // project.rs:289-291 – doppelte Item-ID oder Kollision über Arten.
      if (itemIds.has(item.id) || allIds.has(item.id)) return null;
      itemIds.add(item.id);
      allIds.add(item.id);
      // project.rs:292-294 – referenzierte Quelle muss existieren.
      if (!sourceIds.has(item.sourceId)) return null;
      // Renderer und UI halten genau eine Textur/Zeile je Quelle und Szene.
      if (sceneSourceIds.has(item.sourceId)) return null;
      sceneSourceIds.add(item.sourceId);
    }
    scenes.push(scene);
  }
  if (scenes.length === 0) return null;
  const activeSceneId = value.activeSceneId as string;
  if (!scenes.some((s) => s.id === activeSceneId)) return null;
  return { version: 1, output, sources, scenes, activeSceneId };
}

/** Bild-PiP rechts unten, wie vom Einrichtungsassistenten verwendet. */
export function pipTransform(output: OutputConfig): Transform {
  const width = Math.round(output.width * 0.35);
  const height = Math.round((width * 9) / 16);
  return {
    x: output.width - width - Math.round(output.width * 0.02),
    y: output.height - height - Math.round(output.height * 0.04),
    width,
    height,
    rotationDegrees: 0,
    cropTop: 0,
    cropRight: 0,
    cropBottom: 0,
    cropLeft: 0,
    opacity: 1,
  };
}
